net/src/p2p_network.rs

```diff
@@ -0,0 +1,494 @@
+//! This module defines the trait for holochain networking / p2p
+//! The goal is to make implementing concrete strucs as simple as possible
+//! so that we can iterate quickly on the API.
+//! All calls will go through one of two api functions, one using json
+//! strings, and the other using binary buffers.
+
+use base64;
+use failure::Error;
+use serde_json;
+
+/// callback function type for api methods transporting json data
+pub type ApiFnJson = Box<FnMut(&str) -> Result<String, Error>>;
+
+/// callback function type for api methods transporting binary data
+pub type ApiFnBin = Box<FnMut(&[u8]) -> Result<Vec<u8>, Error>>;
+
+/// when the network is requesting we store data
+/// this callback will be invoked, expecting the data to be validated
+pub type DhtHoldCallback = Box<FnMut(&str) -> Result<bool, Error>>;
+
+/// the identifier for an application
+pub type GenomeHash = [u8; 32];
```

Comment:
Just flagging it here for @thedavidmeister and @lucksus and @ddd-mtl that currently we are just switching to the term `Address` for the content addressable store and the assumption is that that's a SHA256 hash.   What ever this ends up being, these two data structures have to match up in the end!

This doesn't stop my approval of this PR.  Just making sure everybody is aware of the issue.